File: python_solutions/1061.lexicographically-smallest-equivalent-string.cpp
// Containers
#include <deque>
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <unordered_map>
#include <unordered_set>
#include <vector>

#include "conf.d/say.h"

using namespace std;
int dirs[5] = {-1, 0, 1, 0, -1};

class UF {
  int *id, cnt, *sz;

public:
  UF(int N) {
    cnt = N;
    id = new int[N], sz = new int[N];
    for (int i = 0; i < N; i++) id[i] = i, sz[i] = 1;
  }

  ~UF() {
    delete[] id;
    delete[] sz;
  }

  int find(int p) {
    if (p != id[p]) id[p] = find(id[p]);
    return id[p];
  }

  void merge(int x, int y) {
    int i = find(x), j = find(y);
    if (i == j) return;
    id[j] = id[i] = min(i, j);
  }
};

// --------------------------------------------------------
class Solution {
public:
  string smallestEquivalentString(string A, string B, string S) {
    UF uf(26);
    for (int i = 0; i < A.size(); i++) uf.merge(A[i] - 'a', B[i] - 'a');
    string res(S.length(), 'a');
    for (int i = 0; i < S.length(); i++) res[i] = uf.find(S[i] - 'a') + 'a';
    return res;
  }
};

int main() {
  Solution s;
  string A = "leetcode", B = "programs", S = "sourcecode";
  say(s.smallestEquivalentString(A, B, S));
}

File: python_solutions/1061.lexicographically-smallest-equivalent-string.py
from collections import Counter, defaultdict, OrderedDict, deque
from bisect import bisect_left, bisect_right
from functools import reduce, lru_cache
from typing import List
import itertools
import math
import heapq
import string
true = True
false = False
MIN, MAX, MOD = -0x3f3f3f3f, 0x3f3f3f3f, 1000000007
from collections import Counter, defaultdict, OrderedDict, deque
from bisect import bisect_left, bisect_right
from functools import reduce, lru_cache
from typing import List
import itertools
import math
import heapq
import string
true = True
false = False
MIN, MAX, MOD = -0x3f3f3f3f, 0x3f3f3f3f, 1000000007


# https://leetcode.com/problems/lexicographically-smallest-equivalent-string/description/
# Medium
class UF:
    def __init__(self, n):
        self.p = list(range(n))
        self.size = [1] * n

    def merge(self, x, y):
        # Merge by size of set
        px, py = self.find(x), self.find(y)
        if px == py: return
        self.p[px] = self.p[py] = min(px, py)

    def find(self, x):
        if x != self.p[x]:
            self.p[x] = self.find(self.p[x])
        return self.p[x]


class Solution:
    def smallestEquivalentString(self, A: str, B: str, S: str) -> str:
        obj = UF(26)
        for i, a in enumerate(A):
            obj.merge(ord(a) - 97, ord(B[i]) - 97)
        return ''.join(chr(97 + obj.find(ord(c) - 97)) for c in S)


A, B, S = "leetcode", "programs", "sourcecode"
print(Solution().smallestEquivalentString(A, B, S))
